### PR TITLE
add new xscreensaver module

### PR DESCRIPTION
**Pull request by lasers**
Hello there. My best pal @neutronst4r and I aim to... Closes https://github.com/ultrabug/py3status/pull/236 

We took time today to modernize this script the best we can. One part about modernizing this script is to replace commands with `self.py3.command_run`... It works fine... kinda.

The expected behaviour for this module is to leave `xscreensaver` running in the background regardless.

If we start `xscreensaver` and later use `i3-msg start`, the process get killed or something. That was not the case with the current commands. Should we keep the current commands or to use `self.py3.command_run` with some (unknown) parameters?

Please have a look. Many thanks from us. 👨‍👦

### CODE REVIEW

**Comment by lasers** [comment]
That PR is likely abandoned. Completing this PR will closes #236.

**Comment by lasers** [comment]
I added the missing placeholder. I have been running this module for a day. Worked okay. I initially took out cache_timeout, but put it back with 60s. Waiting to be reviewed.

If we don't hear from original PR-er in a week, then we should close that PR and merge this in. I also added right-click to lock screen so this PR is a module enhancement too.

**Comment by neutronst4r** [comment]
I am sorry for the delayed answer. I have had to study hard for finals. I have closed [#236](https://github.com/ultrabug/py3status/pull/236). And I will take a look at your proposed change.

**Comment by neutronst4r** [comment]
Okay, I took a quick look at what you want. Using `self.py3.command_run` and having any program start with a different group id is **not** possible. (This is what you need to launch the program independent of the calling program so the called program doesn't close when the mother program is killed.) That is why I included `preexec_fn=setpgrp` in my original draft.

Without an option to pass aditional `Popen` arguments to `self.py3.command_run` it is not possible. This function could easily be added, but I don't know if it will break other modules and if the other people here even want it.

**Comment by tobes** [comment]
> Without an option to pass aditional Popen arguments to self.py3.command_run it is not possible. 

True, but adding extra options is always possible.  Personally I'm sort of against giving complete free reign but adding a `run_independently` option that would allow this feels like maybe a way forward.

I'm quite keen to have this sort of stuff done by py3 helpers so that we can give a consistent approach eg `command_output()` returns unicode (preventing common errors) also we have an issue around Popen sometimes returning an error code of -15.

Of course this can be done later.

File: py3status/modules/xscreensaver.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    cache_timeout = 60
+    format = '{output}'
+    string_off = u'🖥️'
+    string_on = u'🖥️'
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: `icon_...` would be more in keeping

also I'm not sure that this unicode char is very well supported.

File: py3status/modules/xscreensaver.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    def xscreensaver(self):
+        """
+        """
+        self.run = self.py3.check_commands(['pidof xscreensaver'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: this is broken it is not how `check_commands()` works and you would do that in `post_config_hook()` anyhow.

File: py3status/modules/xscreensaver.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                'color': self.py3.COLOR_BAD
+            }
+
+        return response
+
+    def on_click(self, event):
+        """
+        """
+        if event['button'] == 1:
+            if self.run:
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: this whole `self.run` stuff feels wrong.  we should just see if `xscreensaver` is running when we need to know - storing it just makes us more likely to have incorrect behaviour.

File: py3status/modules/xscreensaver.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        """
+        self.run = self.py3.check_commands(['pidof xscreensaver'])
+        response = {
+            'cached_until': self.py3.time_in(self.cache_timeout),
+            'full_text': self.string_on,
+            'color': self.py3.COLOR_GOOD
+        }
+
+        try:
+            self.run = check_output(['pidof', 'xscreensaver'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: this can still use `py3.command_output()` though you might need to change the except

File: py3status/modules/xscreensaver.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+This script is useful for people who let Xscreensaver manage dpms settings.
+Xscreensaver has its own dpms variables separate from xset. Dpms can be
+safely turned off in xset as long as xscreensaver is running.
+Settings can be managed using "xscreensaver-demo".
+
+Configuration parameters:
+    cache_timeout: refresh interval for this module (default 60)
+    format: default format for this module (default '{output}')
+    string_off: display when Xscreensaver is not running (defaults '🖥️')
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: default~~s~~

File: py3status/modules/xscreensaver.py (hunk 3 above)
Comment (lasers) [inline py3status/modules/xscreensaver.py]: New change. I feel I did it wrong.

**Comment by lasers** [comment]
>True, but adding extra options is always possible. Personally I'm sort of against giving complete free reign but adding a run_independently option that would allow this feels like maybe a way forward.

Should I just wait?

>this is broken it is not how check_commands() works and you would do that in post_config_hook() anyhow.

Fixed.

>icon_... would be more in keeping

Hold for pending discussion. Wanted to fix the bad code first (if any).

>also I'm not sure that this unicode char is very well supported.
FIxed. Now using `SCR` -- or you prefer `XSCR`? Blah.

>this can still use py3.command_output() though you might need to change the except

Fixed. 

I cleaned up some, but I doubt I went around with this well.

**Comment by neutronst4r** [comment]
> Hold for pending discussion. Wanted to fix the bad code first (if any).
It really doesn't matter at all, but if you really want to use the little monitor symbol, I suggest you use [this one](http://fontawesome.io/icon/desktop/), AwesomeFont is much more widespread.

**Comment by tobes** [comment]
>  AwesomeFont is much more widespread.

True, but I think it is important to remember that i3 is used by people for a variety of reasons and one of those is that it is lightweight.  So I think we want to be careful about adding dependencies.

@neutronst4r good to see you about again

File: py3status/modules/xscreensaver.py [hunk 6]
```diff
@@ -11,45 +11,44 @@
 Settings can be managed using "xscreensaver-demo".
 
 Configuration parameters:
-    cache_timeout: refresh interval for this module (default 60)
     format: default format for this module (default '{output}')
-    string_off: display when Xscreensaver is not running (defaults '🖥️')
-    string_on: display when Xscreensaver is running (defaults '🖥️')
+    format_off: display when Xscreensaver is not running (default '')
+    format_on: display when Xscreensaver is running (default '')
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: `icon_..` please 

There are only so many times I can repeat myself.

File: py3status/modules/xscreensaver.py [hunk 7]
```diff
@@ -11,45 +11,44 @@
...
 from os import setpgrp
-from subprocess import Popen, DEVNULL, check_output, CalledProcessError
+from subprocess import Popen, DEVNULL
 
 
 class Py3status:
     """
     """
     # available configuration parameters
-    cache_timeout = 60
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: why did you remove this? it allows user control and is pretty standard, you are now killing the cpu.

File: py3status/modules/xscreensaver.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    def xscreensaver(self):
+        """
+        """
+        response = {
+            'full_text': self.format_on,
+            'color': self.py3.COLOR_GOOD
+        }
+
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: why set `self.run`?  there is no need to store this.  You just care about the Exception

File: py3status/modules/xscreensaver.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    def on_click(self, event):
+        """
+        """
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
+        except:
+            self.run = False
+
+        if event['button'] == 1:
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: I quite like having `button_toggle`  `button_lock_screen` in the config then you can just do 

`if event['button'] == self.button_toggle:` etc and the user can choose the button and or disable the action. see module clock for example

File: py3status/modules/xscreensaver.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            if self.run:
+                self.run = Popen(['killall',
+                                  'xscreensaver'],
+                                 stdout=DEVNULL,
+                                 stderr=DEVNULL,
+                                 preexec_fn=setpgrp) == 0
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: This does not need `preexec_fn=setpgrp`  just use `self.py3.command_run('killall xscreensaver')` and ignore the result.

File: py3status/modules/xscreensaver.py (hunk 7 above)
Comment (lasers) [inline py3status/modules/xscreensaver.py]: Does this mean dpms should be fixed too? It's missing a `cache_timeout` too.
That's where I got the clue from.

I was trying to avoid this https://github.com/ultrabug/py3status/issues/580 by optimizing a bit. Getting it to not run something every ___ seconds, but to be updated only on click.

I think if somebody is planning on using this xscreensaver module, then they would be only xscreensaver to stop or run xscreensaver... basically CACHE_FOREVER.

It'll be back.

**Comment by lasers** [comment]
@neutronst4r Thanks. I like that icon. It's much better. It fits well with the monitor from xscreensaver... https://xubuntu.org/wp-content/uploads/2014/03/xubuntu_lock_xscreensaver.png

I pushed the changes. Get with the times a bit. No more CRT monitor.

@tobes All fixed? I put `cache_timeout` back in.. w/ 15s, add `button_toggle` and `button_lock`, replaced `format_` to `icon_`. I want to keep font-awesome icon. It's more clean than `SCR` or the old icon. I hope it's ready. 👍

File: py3status/modules/xscreensaver.py (hunk 6 above)
Comment (lasers) [inline py3status/modules/xscreensaver.py]: It was done immediately / silently. I didn't see what you said about dependencies until several hours later. `SCR`? Will look for square box.

File: py3status/modules/xscreensaver.py (hunk 7 above)
Comment (tobes) [inline py3status/modules/xscreensaver.py]: Actually it's not updating quite as much as I feared.  But, yes we want to check.  The user may have started stopped the xscreensaver independently etc.

Thanks

File: py3status/modules/xscreensaver.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                'color': self.py3.COLOR_BAD
+            }
+
+        return response
+
+    def on_click(self, event):
+        """
+        """
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: can `self.run` just be `run`  by using `self.` we are storing this in the object.  Seeing as we do not use it outside of this function it is misleading and makes the code much harder to follow.  Here it is quite obvious but in a more complex module it would be nasty and leads to errors.  It is also a maintenance overhead.

File: py3status/modules/xscreensaver.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        try:
+            self.py3.command_output(['pidof', 'xscreensaver'])
+        except:
+            response = {
+                'cached_until': self.py3.time_in(self.cache_timeout),
+                'full_text': self.icon_off,
+                'color': self.py3.COLOR_BAD
+            }
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: can we just set `icon` and `color` and then create the response once.

Currently if there is no xscreensaver running we are calling `time_in()` and `safe_format()` twice which is a lot of unneeded work.  We want each module to be a light as it can be.

File: py3status/modules/xscreensaver.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        if event['button'] == self.button_lock:
+            Popen(['xscreensaver-command', '-lock'],
+                  stdout=DEVNULL, stderr=DEVNULL, preexec_fn=setpgrp) == 1
+
+        if event['button'] == self.button_toggle:
+            if self.run:
+                self.py3.command_run(['killall', 'xscreensaver'])
+            else:
+                Popen(['xscreensaver', '-no-splash', '-no-capture-stderr'],
+                      stdout=DEVNULL, stderr=DEVNULL, preexec_fn=setpgrp) == 1
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: can we add a comment like
```
# Because we want xscreensaver to continue running even if py3status exits
# we need to run using preexec_fn=setpgrp
```

we don't need the ` == 1` test I think

File: py3status/modules/xscreensaver.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        """
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
+        except:
+            self.run = False
+
+        if event['button'] == self.button_lock:
+            Popen(['xscreensaver-command', '-lock'],
+                  stdout=DEVNULL, stderr=DEVNULL, preexec_fn=setpgrp) == 1
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: Three things.

1) this does not need preexec_fn=setpgrp so we can use `py3.command_run()`

2) we don't need the `== `1 test

3) this only works if xsreensaver is running?  So should we not check that and start xscreensaver if needed and then lock it?

File: py3status/modules/xscreensaver.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+        if event['button'] == self.button_toggle:
+            if self.run:
+                self.py3.command_run(['killall', 'xscreensaver'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: whilst reading the `xscreensaver-command` man page I saw this.

>-exit

>Causes the xscreensaver process to exit gracefully. This does nothing if the display is currently locked.
Warning: never use kill -9 with xscreensaver while the screensaver is active. If you are using a virtual root window manager, that can leave things in an inconsistent state, and you may need to restart your window manager to repair the damage. 

So maybe we should use this recommended method.

**Comment by lasers** [comment]
> The review that just keeps giving

Thank you. Reviews matters.  I know it can be annoying if I made mistakes or got things done differently than you prefer. I wanted to get things done fast whilst keeping variables simple (but more importantly, things are being done correctly too).

Other issues. I will look at them and make changes in few hours.
I mark Janurary 26 as Review Appreciation Day. Cheers. 🏆

**Comment by lasers** [comment]
> can we just set icon and color and then create the response once.

Fixed.

> Currently if there is no xscreensaver running we are calling time_in() and safe_format() twice which is a lot of unneeded work. We want each module to be a light as it can be.

Not sure what I should do here.

> can self.run just be run by using self.

Fixed. I followed dpms.


> this only works if xsreensaver is running? So should we not check that and start xscreensaver if needed and then lock it?

No running = No locking. Run it first (imho).

> -exit // Causes the xscreensaver process to exit gracefully

I tried that before. Not working. Working now tho. Maybe I used wrong py3.command. :confused:

> can we add a comment like

Yes sir. Done.

**Comment by tobes** [comment]
> No running = No locking. Run it first (imho).

I'd say so :)

File: py3status/modules/xscreensaver.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    button_toggle: mouse button to toggle Xscreensaver (default 1)
+    cache_timeout: refresh interval for this module (default 15)
+    format: default format for this module (default '{icon}')
+    icon_off: display when Xscreensaver is not running (default 'XSCR')
+    icon_on: display when Xscreensaver is running (default 'XSCR')
+
+ Format placeholders:
+    {icon} display current xscreensaver icon
+
+@author neutronst4r <c7420{at}gmx{dot}net>
```
Comment (neutronst4r) [inline py3status/modules/xscreensaver.py]: change email to `<c7420{at}posteo{dot}net>` please and add yourself as an author, if you like

File: py3status/modules/xscreensaver.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+            self.py3.notify_user('xscreensaver is not installed. cannot use this module.')
+
+    def xscreensaver(self):
+        """
+        """
+        try:
+            run = self.py3.command_output(['pidof', 'xscreensaver'])
+        except:
+            run = False
+
```
Comment (neutronst4r) [inline py3status/modules/xscreensaver.py]: As @tobes has suggested saving the result of run is kind of unnecessary. We could do something like:
```
def _is_running_(self):
       """
       tests whether xscreensaver is running, returns boolean
       """
       try:
           self.py3.command_output(['pidof', 'xscreensaver'])
       except:
           return False
       else:
           return True
```
and ask if xscreensaver is running.

Comment (lasers) [inline py3status/modules/xscreensaver.py]: Thanks. I didn't understand that. I jumped in here to muck together a wanted feature for `arch_updates` to stay hidden if none... then I keep going with many other little things.

I fixed your email and am not worthy of a mention on any script. I was wondering how to make a function so I don't do two same things in different places. This has been updated now. Cheers.

File: py3status/modules/xscreensaver.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        else:
+            return True
+
+    def xscreensaver(self):
+        if not self.py3.check_commands(['xscreensaver']):
+            return {
+                'cached_until': self.py3.CACHE_FOREVER,
+                'color': self.py3.COLOR_BAD,
+                'full_text': STRING_UNAVAILABLE
+            }
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: This should be done in `post_config_hook()` seeing as we have it defined.  By raising an `Exception(msg)` less lines too

File: py3status/modules/xscreensaver.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        return {
+            'cached_until': self.py3.time_in(self.cache_timeout),
+            'full_text': self.py3.safe_format(self.format, {'icon': icon}),
+            'color': color
+        }
+
+    def on_click(self, event):
+        run = self._is_running()
+        if event['button'] == self.button_lock:
+            self.py3.command_run(['xscreensaver-command', '-lock'])
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: should we not use `xscreensaver-command -activate`?  If I have setup my xscreensaver to not lock the screen then I would not expect it to be locked here.

File: py3status/modules/xscreensaver.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    button_toggle: mouse button to toggle Xscreensaver (default 1)
+    cache_timeout: refresh interval for this module (default 15)
+    format: default format for this module (default '{icon}')
+    icon_off: display when Xscreensaver is not running (default 'XSCR')
+    icon_on: display when Xscreensaver is running (default 'XSCR')
+
+ Format placeholders:
+    {icon} Xscreensaver icon
+
+@author neutronst4r <c7420{at}posteo{dot}net>
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: I don't think you should add someone else as the author also if this is based on dpms then neutronst4r is not the author of that module.

File: py3status/modules/xscreensaver.py [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+"""
+from os import setpgrp
+from subprocess import Popen, DEVNULL
+STRING_UNAVAILABLE = "xscreensaver: isn't installed"
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: A blank line would be nice here

File: py3status/modules/xscreensaver.py (hunk 20 above)
Comment (tobes) [inline py3status/modules/xscreensaver.py]: actually read your initial comment here so maybe
`@author neutronst4r, lasers`

**Comment by lasers** [comment]
Hi there.
* You want blank lines? You got them.
* Exception now in `psc` using `raise Exception(msg)` to use that new/nice exception string in py3status instead of old annoying i3-nagbar exceptions.
* I tested `-activate`. Yes. Good command. Can configure `-lock` using `xscreensaver-demo` (graphical) or edit the configuration file `~/.xscreensaver`.

File: py3status/modules/xscreensaver.py [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+from subprocess import Popen, PIPE
+
+STRING_UNAVAILABLE = "isn't installed"
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    button_lock = 3
```
Comment (tobes) [inline py3status/modules/xscreensaver.py]: maybe we should rename this to `button_activate` to match it's function

Comment (lasers) [inline py3status/modules/xscreensaver.py]: Yup. Done.

File: py3status/modules/xscreensaver.py (hunk 20 above)
Comment (lasers) [inline py3status/modules/xscreensaver.py]: Blah. Done. Let's tag it ready to go and get out of here.

**Comment by tobes** [comment]
Thanks @lasers this is now good to go as far as I'm concerned.

**Comment by ultrabug** [comment]
Okay I guess the label is good, thanks to everyone !
